SmartDeviceLink/SDLCarWindow.m

```diff
@@ -0,0 +1,257 @@
+
+@property (weak, nonatomic, nullable) SDLStreamingMediaLifecycleManager *streamManager;
+
+@property (assign, nonatomic, getter=isLockScreenMoving) BOOL lockScreenMoving;
+
+@end
+
+@implementation SDLCarWindow
+
+- (instancetype)initWithStreamManager:(SDLStreamingMediaLifecycleManager *)streamManager targetFramerate:(NSUInteger)framesPerSecond {
+    self = [super init];
+    if (!self) { return nil; }
+
+    SDLLogD(@"Initializing Car Window automatic streaming with framerate: %lu", (unsigned long)framesPerSecond);
+
+    _streamManager = streamManager;
+    _targetFramerate = framesPerSecond;
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didReceiveVideoStreamStarted:) name:SDLVideoStreamDidStartNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_didReceiveVideoStreamStopped:) name:SDLVideoStreamDidStopNotification object:nil];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_lockScreenMoving:) name:SDLLockScreenManagerWillPresentLockScreenViewController object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_lockScreenMoving:) name:SDLLockScreenManagerWillDismissLockScreenViewController object:nil];
+
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_lockScreenStoppedMoving:) name:SDLLockScreenManagerDidPresentLockScreenViewController object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_lockScreenStoppedMoving:) name:SDLLockScreenManagerDidDismissLockScreenViewController object:nil];
+
+    return self;
+}
+
+- (void)sdl_sendFrame:(CADisplayLink *)displayLink {
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (!self.streamManager.isVideoConnected || self.streamManager.isVideoStreamingPaused) {
+            return;
+        }
+        
+        if (self.sameFrameCounter == 30 && ((displayLink.timestamp - self.lastMd5HashTimestamp) <= 0.1)) {
+            SDLLogD(@"Paused CarWindow, no frame changes in over a second");
+            return;
+        }
+
+        if (self.isLockScreenMoving) {
+            SDLLogD(@"Paused CarWindow, lock screen moving");
+            return;
+        }
+        
+        self.lastMd5HashTimestamp = displayLink.timestamp;
+        
+        CGRect bounds = self.rootViewController.view.bounds;
+        
+        UIGraphicsBeginImageContextWithOptions(bounds.size, YES, 1.0f);
+        [self.rootViewController.view drawViewHierarchyInRect:bounds afterScreenUpdates:YES];
+        UIImage *screenshot = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        
+        CGImageRef imageRef = screenshot.CGImage;
+        
+        // We use MD5 Hashes to determine if we are sending the same frame over and over. If so, we will only send 30.
+        NSString *currentMd5Hash = [self.class sdl_md5HashForImageRef:imageRef];
```

Comment:
I think this is a carry-over from the Ford prototype. Was there any testing to see if the md5 hashing limits the framerate at all?

Reply:
Yes, when the application is static on the screen (no changes at all, which, admittedly, is not very often) it cuts the CPU usage in half.

Reply:
Understood. I was just curious if when the application is trying to stream at 30+ fps, if hashing each frame began to become a bottleneck.

Reply:
No, definitely not, virtually all of the CPU is used in the screenshotting.